File: Silver IV/BOJ_10866/BOJ_10866.cpp
#include <deque>
#include <iostream>
#include <string>

using namespace std;

int n, num;
string str;
deque<int> dq;

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n;
    while (n--) {
        cin >> str;

        if (str == "push_front") {
            cin >> num;
            dq.push_front(num);
        }
        else if (str == "push_back") {
            cin >> num;
            dq.push_back(num);
        }
        else if (str == "pop_front") {
            if (dq.empty()) {
                cout << "-1" << '\n';
            }
            else {
                cout << dq.front() << '\n';
                dq.pop_front();
            }
        }
        else if (str == "pop_back") {
            if (dq.empty()) {
                cout << "-1" << '\n';
            }
            else {
                cout << dq.back() << '\n';
                dq.pop_back();
            }
        }
        else if (str == "size") {
            cout << dq.size() << '\n';
        }
        else if (str == "empty") {
            if (dq.empty()) {
                cout << '1' << '\n';
            }
            else {
                cout << '0' << '\n';
            }
        }
        else if (str == "front") {
            if (dq.empty()) {
                cout << "-1" << '\n';
            }
            else {
                cout << dq.front() << '\n';
            }
        }
        // str == "back"
        else {
            if (dq.empty()) {
                cout << "-1" << '\n';
            }
            else {
                cout << dq.back() << '\n';
            }
        }
    }

    return 0;
}

File: Silver IV/BOJ_10866/BOJ_10866.py
import sys
from collections import deque

if __name__ == "__main__":
    dq = deque()

    n = int(sys.stdin.readline())
    info = [sys.stdin.readline().strip() for _ in range(n)]

    # print(info)

    for row in info:
        if row == "size":
            print(len(dq))
        elif row == "back":
            if dq:
                value = dq.pop()
                print(value)
                dq.append(value)
            else:
                print(-1)
        elif row == "empty":
            if dq:
                print(0)
            else:
                print(1)
        elif row == "front":
            if dq:
                value = dq.popleft()
                print(value)
                dq.appendleft(value)
            else:
                print(-1)
        elif row == "pop_back":
            if dq:
                print(dq.pop())
            else:
                print(-1)
        elif row == "pop_front":
            if dq:
                print(dq.popleft())
            else:
                print(-1)
        elif row[0:9] == "push_back":
            dq.append(int(row[10:]))
        elif row[0:10] == "push_front":
            dq.appendleft(int(row[11:]))
